### PR TITLE
free function serialization DSL

### PR DESCRIPTION
**Pull request by jeffro256**
Useful for PR #25

### CODE REVIEW

**Comment by jeffro256** [comment]
Example usage (replacing serialization_demo_utils):
```
BLOB_SERIALIZER(sp::jamtis::address_index_t);
BLOB_SERIALIZER(sp::jamtis::address_tag_t);
BLOB_SERIALIZER(sp::jamtis::encrypted_address_tag_t);
BLOB_SERIALIZER(sp::jamtis::encoded_amount_t);

BEGIN_SERIALIZE_OBJECT_FN(sp::SpCoinbaseEnoteCore)
    FIELD_F(onetime_address)
    VARINT_FIELD_F(amount)
END_SERIALIZE()

BEGIN_SERIALIZE_OBJECT_FN(sp::SpEnoteCore)
    FIELD_F(onetime_address)
    FIELD_F(amount_commitment)
END_SERIALIZE()

BEGIN_SERIALIZE_OBJECT_FN(sp::SpEnoteImageCore)
    FIELD_F(masked_address)
    FIELD_F(masked_commitment)
    FIELD_F(key_image)
END_SERIALIZE()

BEGIN_SERIALIZE_OBJECT_FN(sp::SpEnoteImageCore)
    FIELD_F(masked_address)
    FIELD_F(masked_commitment)
    FIELD_F(key_image)
END_SERIALIZE()

BEGIN_SERIALIZE_OBJECT_FN(sp::BulletproofPlus2)
    FIELD_F(A)
    FIELD_F(A1)
    FIELD_F(B)
    FIELD_F(r1)
    FIELD_F(s1)
    FIELD_F(d1)
    FIELD_F(L)
    FIELD_F(R)
END_SERIALIZE()

... continue pattern here 
```

We can serialize directly from seraphis types non-intrusively and don't need all the extra types and boilerplate eventually

**Comment by DangerousFreedom1984** [comment]
That should make things much easier. Thanks.

**Comment by UkoeHB** [comment]
You should parallel-PR this to the monero repo.

File: src/serialization/serialization.h [hunk 1]
```diff
@@ -123,6 +123,16 @@ inline bool do_serialize(Archive &ar, bool &v)
   template <bool W, template <bool> class Archive>			\
   bool member_do_serialize(Archive<W> &ar) {
 
+/*! \macro BEGIN_SERIALIZE_FN
+ *
+ * \brief Begins the environment of the DSL as a free function
+ * \detailed for describing how to
+ * serialize an of an archive type
```
Comment (UkoeHB) [inline src/serialization/serialization.h]: Fix this comment

File: src/serialization/serialization.h [hunk 2]
```diff
@@ -139,6 +149,25 @@ inline bool do_serialize(Archive &ar, bool &v)
   template <bool W, template <bool> class Archive>			\
   bool do_serialize_object(Archive<W> &ar){
 
+/*! \macro BEGIN_SERIALIZE_OBJECT_FN
+ *
+ * \brief Begins the environment of the DSL as a free function in object-style
+ * \detailed for describing how to
+ * serialize an of an archive type
```
Comment (UkoeHB) [inline src/serialization/serialization.h]: Fix this comment

File: src/serialization/serialization.h [hunk 3]
```diff
@@ -173,6 +202,12 @@ inline bool do_serialize(Archive &ar, bool &v)
     if (!r || !ar.good()) return false;			\
   } while(0);
 
+/*! \macro FIELD_F(f)
+ *
+ * \brief tags the field with the variable name and then serializes it (for use in free function)
```
Comment (UkoeHB) [inline src/serialization/serialization.h]:
```suggestion
 * \brief tags the field with the variable name and then serializes it (for use in a free function)
```

**Review by UkoeHB — CHANGES_REQUESTED** [review]
@vtnerd can you review this?

File: src/serialization/serialization.h [hunk 4]
```diff
@@ -173,6 +202,12 @@ inline bool do_serialize(Archive &ar, bool &v)
...
+ */
+#define FIELD_F(f) FIELD_N(#f, v.f)
```
Comment (vtnerd) [inline src/serialization/serialization.h:212]: I don't understand what this does, other than it allows the macro to be used outside of a class scope. Presumably that is the goal here?

File: src/serialization/serialization.h [hunk 5]
```diff
@@ -139,6 +149,25 @@ inline bool do_serialize(Archive &ar, bool &v)
   template <bool W, template <bool> class Archive>			\
   bool do_serialize_object(Archive<W> &ar){
 
+/*! \macro BEGIN_SERIALIZE_OBJECT_FN
+ *
+ * \brief Begins the environment of the DSL as a free function in object-style
+ * \detailed for describing how to
+ * serialize an of an archive type
+ */
+#define BEGIN_SERIALIZE_OBJECT_FN(stype)               \
```
Comment (vtnerd) [inline src/serialization/serialization.h:161]: I agree that the comment needs some fixing - the goal is to allow a standalone function in a cpp or something? Difficult to say without any example uses.

File: src/serialization/serialization.h (hunk 4 above)
Comment (jeffro256) [reply src/serialization/serialization.h:212]: Yes, that's the current design goal with the Seraphis lib AFAIK: separate class definitions and their serialization.

File: src/serialization/serialization.h (hunk 5 above)
Comment (jeffro256) [reply src/serialization/serialization.h:161]: There's some sample uses in an earlier comment in this thread. Should I put it into the code?

**Comment by jeffro256** [comment]
Fixed comments, thanks @UkoeHB

**Review by UkoeHB — APPROVED** [review]

**Pull request merged**